File: SeaCleaner/Client/Game/Game.cs
using Microsoft.JSInterop;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace SeaCleaner.Client.Game
{
    internal enum GameState { GAME_PLAY, GAME_PAUSE, GAME_LOST, GAME_WON }

    internal class Game
    {
        public const int SCREEN_WIDTH = 1000;
        public const int SCREEN_HEIGHT = 700;
        public const int UPDATE_INTERVAL = 5;
        public const int DEPTH_Y_POS = 190;
        public const int DEPTH_HEIGHT = 510;
        public const int WAVES_Y_POS = 120;
        public const bool SHOW_BB = false;

        private readonly IJSRuntime _jsRuntime;
        private readonly string _canvasId;
        public Dictionary<string, SpriteImageInfo> Sprites;
        private Timer _timer;
        private DateTimeOffset _startTime;
        private readonly Random _random = new Random();

        public GameState State { get; private set; }

        private Sprite _logo;
        private Sprite _seaBack;
        private Sprite _seaFront;
        private Sprite _corrals;
        private Sprite _platePause;
        private Sprite _plateWon;
        private Sprite _plateLost;
        private Waves _waves;

        private readonly List<Bubble> _bubbles = new List<Bubble>();
        private readonly List<Fishes> _fishes = new List<Fishes>();
        public List<Trash> Trashes { get; } = new List<Trash>();
        private int _activeFishesIndex1 = 0;
        private int _activeFishesIndex2 = 5;

        private readonly List<Dolphin> _dolphins = new List<Dolphin>();
        private Ship _ship;        

        public Game(IJSRuntime jsRuntime, string canvasId)
        {
            _jsRuntime = jsRuntime;
            _canvasId = canvasId;
            State = GameState.GAME_PLAY;
        }

        public async ValueTask LoadResources()
        {
            Sprites = await GameResources.LoadImages(_jsRuntime);
        }

        public async ValueTask Initialize()
        {
            await _jsRuntime.InvokeVoidAsync("initializeScreen", _canvasId);
            _logo = new Sprite(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT, Sprites["Logo"]);
            await _logo.Draw(_jsRuntime);

            _seaBack = new Sprite(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT, Sprites["SeaBack"]);
            _seaFront = new Sprite(0, DEPTH_Y_POS, SCREEN_WIDTH, SCREEN_HEIGHT, Sprites["SeaFront"]);
            _corrals = new Sprite(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT, Sprites["Corrals"]);
            _platePause = new Sprite(294, -420, 413, 413, Sprites["PlatePause"]);
            _plateWon = new Sprite(294, -420, 413, 413, Sprites["PlateWon"]);
            _plateLost = new Sprite(294, -420, 413, 413, Sprites["PlateLost"]);

            _waves = new Waves(WAVES_Y_POS, 15, SCREEN_WIDTH + 100, 5, Sprites["WavesFront"], Sprites["WavesBack"]);

            for (int i = 0; i < 20; i++)
            {
                _bubbles.Add(new Bubble(_random.NextDouble() * SCREEN_WIDTH, _random.NextDouble() * SCREEN_WIDTH % 550, 5, -2, 1, Sprites["Bubble"]));
            }

            bool b = false;
            string[] fNames = new[] { "Fish1L", "Fish1R", "Fish2L", "Fish2R", "Fish3L", "Fish3R" };
            for (int i = 0; i < fNames.Length; i++)
            {
                _fishes.Add(new Fishes(5, 5, b = !b, Sprites[fNames[i]]));
                _fishes.Last().Initialize();
            }

            string[] tNames = new[] { "Trash1", "Trash2", "Trash3", "Trash4", "Trash5" };
            for (int i = 0; i < 5; i++)
            {
                Trashes.Add(new Trash(Sprites[tNames[i]]));
                Trashes.Add(new Trash(Sprites[tNames[i]]));
            }
            _trashesCount = Trashes.Count;

            b = false;
            Dolphin._dPos = 0;
            for (int i = 0; i < 4; i++)
            {
                b = !b;
                var d = new Dolphin(b,
                    Sprites["DolphinFlow" + (b ? "L" : "R")],
                    Sprites["DolphinEat" + (b ? "L" : "R")],
                    Sprites["DolphinDie" + (b ? "L" : "R")],
                    CheckLost, CheckWon);
                d.Initialize();
                _dolphins.Add(d);
            }
            _dolphinsCount = _dolphins.Count;

            _ship = new Ship(this);
        }

        public async ValueTask Update()
        {
            if (State == GameState.GAME_PAUSE && _platePause.PosY >= 0) return;

            if (State == GameState.GAME_LOST)
            {
                if (_plateLost.PosY < 0)
                    _plateLost.PosY += 30;
                else
                    await EndGame();
                return;
            }

            if (State == GameState.GAME_WON)
            {
                if (_plateWon.PosY < 0)
                    _plateWon.PosY += 30;
                else
                    await EndGame();
                return;
            }

            if (State == GameState.GAME_PAUSE)
            {
                if (_platePause.PosY < 0)
                    _platePause.PosY += 30;
                return;
            }

            if (State == GameState.GAME_PLAY && _platePause.PosY > -420)
            {
                _platePause.PosY -= 30;
                return;
            }

            _waves.Update();

            for (int i = 0; i < 20; i++) _bubbles[i].Update();

            _fishes[_activeFishesIndex1].Update();
            _fishes[_activeFishesIndex2].Update();

            if (_fishes[_activeFishesIndex1].IsOffScreen())
            {
                _fishes[_activeFishesIndex1].Initialize();
                _activeFishesIndex1 = _random.Next(0, 6);
                if (_activeFishesIndex1 == _activeFishesIndex2) _activeFishesIndex1 = (_activeFishesIndex1 + 1) % 6;
            }

            if (_fishes[_activeFishesIndex2].IsOffScreen())
            {
                _fishes[_activeFishesIndex2].Initialize();
                _activeFishesIndex2 = _random.Next(0, 6);
                if (_activeFishesIndex2 == _activeFishesIndex1) _activeFishesIndex2 = (_activeFishesIndex2 + 1) % 6;
            }

            foreach(var t in Trashes)
            {
                t.Update();
            }

            foreach (var d in _dolphins)
            {
                d.Update();

                if (d.State == DolphinState.Flow)
                {
                    foreach (var t in Trashes)
                    {
                        if (t.IsCaptured) continue;

                        if (d.MouthBB.IsIntersect(t.BBox))
                        {
                            t.IsCaptured = true;
                            d.CapturedTrash = t;
                            d.State = DolphinState.Eat;
                            break;
                        }
                    }
                }
            }

            foreach(var d in _dolphins)
            {
                if(d.State == DolphinState.Flow && d.BodyBB.IsIntersect(_ship.Arrow.Hug.HugOuterBBox))
                {
                    d.State = DolphinState.Die;
                }
            }

            _ship.Update();
        }

        public async ValueTask Draw()
        {
            if (State == GameState.GAME_PAUSE && _platePause.PosY >= 0) return;

            await _seaBack.Draw(_jsRuntime);
            await _waves.DrawBack(_jsRuntime);

            await _fishes[_activeFishesIndex1].Draw(_jsRuntime);
            await _fishes[_activeFishesIndex2].Draw(_jsRuntime);
            await _corrals.Draw(_jsRuntime);
            foreach(var t in Trashes)
            {
                await t.Draw(_jsRuntime);
            }

            foreach(var d in _dolphins)
            {
                await d.Draw(_jsRuntime);
            }

            await _ship.Draw(_jsRuntime);

            foreach (var b in _bubbles)
            {
                await b.Draw(_jsRuntime);
            }

            await _seaFront.Draw(_jsRuntime);
            await _waves.DrawFront(_jsRuntime);

            switch (State)
            {
                case GameState.GAME_PLAY:
                    if (_platePause.PosY > -420) await _platePause.Draw(_jsRuntime);
                    break;
                case GameState.GAME_PAUSE:
                    await _platePause.Draw(_jsRuntime);
                    break;
                case GameState.GAME_WON:
                    await _plateWon.Draw(_jsRuntime);
                    break;
                case GameState.GAME_LOST:
                    await _plateLost.Draw(_jsRuntime);
                    break;
            }
        }

        public async ValueTask MainLoop()
        {
            await Update();
            await Draw();
        }

        [JSInvokable]
        public void OnKeyDown(int keyCode)
        {
            if(keyCode == 80)
            {
                if (State == GameState.GAME_PLAY)
                    State = GameState.GAME_PAUSE;
                else
                    if (State == GameState.GAME_PAUSE)
                        State = GameState.GAME_PLAY;
            }

            _ship.KeyDown(keyCode);
        }

        [JSInvokable]
        public void OnKeyUp(int keyCode)
        {
            _ship.KeyUp(keyCode);
        }

        private Action<int> _onDolphinDie;
        private Action<int> _onTrashCatched;
        private Func<bool, int, long, ValueTask> _onGameEnd;
        private Action _afterGameFinished;

        private DotNetObjectReference<Game> _gameDotNetReference;

        public async ValueTask StartGame(Action<int> onDolphinDie, Action<int> onTrashCatched, Func<bool, int, long, ValueTask> onGameEnd, Action afterGameFinished)
        {
            _onDolphinDie = onDolphinDie;
            _onTrashCatched = onTrashCatched;
            _onGameEnd = onGameEnd;
            _afterGameFinished = afterGameFinished;

            _gameDotNetReference = DotNetObjectReference.Create(this);
            await _jsRuntime.InvokeVoidAsync("setupKeyEvents", _gameDotNetReference);

            _startTime = DateTimeOffset.Now;
            _timer = new Timer(async _ => await MainLoop(), null, 500, UPDATE_INTERVAL);
        }

        private async ValueTask EndGame()
        {
            if(_timer != null) await _timer.DisposeAsync();
            await _jsRuntime.InvokeVoidAsync("clearKeyEvents");
            await _jsRuntime.InvokeVoidAsync("clearResources");

            Thread.Sleep(1500);
            
            _gameDotNetReference?.Dispose();            

            int dCount = _dolphins.Count(d => d.State == DolphinState.Flow);
            long seconds = Convert.ToInt64((DateTimeOffset.Now - _startTime).TotalSeconds);
            bool victory = dCount > 0;
            await _onGameEnd(victory, dCount, seconds);            

            _afterGameFinished();
        }

        private int _dolphinsCount;
        private int _trashesCount;

        public void CheckLost()
        {
            int c = _dolphins.Count(d => d.State == DolphinState.Flow);
            if(c < _dolphinsCount)
            {
                _dolphinsCount = c;
                _onDolphinDie(c);
            }

            if(c == 0)
            {
                State = GameState.GAME_LOST;
            }
        }

        public void CheckWon()
        {
            int c = Trashes.Count(t => !t.IsCaptured);
            if(c < _trashesCount)
            {
                _trashesCount = c;
                _onTrashCatched(c);
            }

            if(c == 0 && _dolphinsCount > 0)
            {
                State = GameState.GAME_WON;
            }
        }
    }
}

File: SeaCleaner/Client/wwwroot/js/Game.js
//
const SCREEN_WIDTH  = 1000;
const SCREEN_HEIGHT = 700;
//
const UPDATE_INTERVAL = 25;
//
const DEPTH_Y_POS = 190; 
const DEPTH_HEIGHT = 510;
const WAVES_Y_POS = 120;
//
const GAME_PLAY = 0;
const GAME_PAUSE = 1;
const GAME_LOST = 2;
const GAME_WON = 3;
//
let Game = {
    logo      : new CSprite(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT),
    waves     : new CWaves(WAVES_Y_POS, 15, SCREEN_WIDTH),
    seaBk     : new CSprite(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT),
    corrals   : new CSprite(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT),
    seaTop    : new CSprite(0, DEPTH_Y_POS, SCREEN_WIDTH, DEPTH_HEIGHT),
    platePause : new CSprite(294, -420, 413, 413),
    plateWon  : new CSprite(294, -420, 413, 413),
    plateLost : new CSprite(294, -420, 413, 413),
    state : GAME_PLAY,    
    bubbles : [],  
    fishes  : [], 
    fishesId : [],
    trashes : [],
    dolphins : [],
    //
    timerId: -1,       
    //
    loadContent : function(){
        //
        this.logo.loadSprite("/images/Logo.jpg");
        //MultiImage.loadImages();
        this.seaBk.loadSprite("/images/sea_bk.jpg");
        this.corrals.loadSprite("/images/corrals.png");
        this.seaTop.loadSprite("/images/sea_top.png");
        this.waves.LoadSprite(5, 7, '/images/wave_bk.png', '/images/wave_top.png');
        this.plateLost.loadSprite("/images/lost.png");
        this.platePause.loadSprite("/images/pause.png");
        this.plateWon.loadSprite("/images/won.png");
        //
        for(let i = 0; i< 20; i++)
            this.bubbles.push(new CBubble(0, 0, 5, -2, 1));
        //
        let b = false;
        let fNames = ['fish_1_l', 'fish_1_r', 'fish_2_l',
                      'fish_2_r', 'fish_3_l', 'fish_3_r'];
        for(let i=0; i< fNames.length; i++)
            this.fishes.push(new CFishes(5, 5, b = !b, fNames[i]));
        //
        let tNames = ['trash_1', 'trash_2', 'trash_3', 'trash_4', 'trash_5'];
        for(let i=0; i< 5; i++){
            this.trashes.push(new CTrash(tNames[i]));
            this.trashes.push(new CTrash(tNames[i]));
        }   
        //
        b = false;
        for (let i =0; i< 4; i++)
            this.dolphins.push(new CDolphin(b = !b));
        //
        Ship.loadContent();        
    },
    //
    initialize  : function(canvas){
        this.context = canvas.getContext('2d');
        this.logo.draw(this.context);
        //this.context.lineWidth = 2;
        //
        for (let i = 0; i < 20; i++){
            this.bubbles[i].PosX = Math.random()*1000;
            this.bubbles[i].PosY = Math.random() * 1000 % 550; /*+ 550*/
        }            
        for (let i = 0; i < 6; i++)
            this.fishes[i].initialize();
        this.fishesId.push(0);
        this.fishesId.push(5);
        //
        for (let i = 0; i < this.trashes.length; i++)
            this.trashes[i].initialize();
        //
        for (let i=0; i< 4; i++)
            this.dolphins[i].initialize();
        //       
        Ship.initialize();             
    },
    //
    update : function(){
        if(this.state === GAME_LOST){
            if (this.plateLost.PosY < 0)
                this.plateLost.PosY += 30;
            else
                this.onGameFinish();
            return;
        }
        if(this.state === GAME_WON){
            if(this.plateWon.PosY < 0)
                this.plateWon.PosY += 30;
            else
                this.onGameFinish();
            return;
        }
        if(this.state === GAME_PAUSE){
            if(this.platePause.PosY < 0)
                this.platePause.PosY += 30;
            return;
        }
        if(this.state === GAME_PLAY && this.platePause.PosY > -420){
            this.platePause.PosY -= 30;
            return;
        }        
        this.waves.Update();
        //
        for (let i=0; i<20; i++)
            this.bubbles[i].update();
        //
        for (let i =0; i< 2; i++)
            this.fishes[this.fishesId[i]].update();
        //
        if(this.fishes[this.fishesId[0]].isOffScreen()){
            this.fishes[this.fishesId[0]].initialize();
            this.fishesId[0] = Math.round(Math.random()*1000) % 6;
            while(this.fishesId[0] === this.fishesId[1])
                this.fishesId[0] = Math.round(Math.random()*1000) % 6;            
        }
        if(this.fishes[this.fishesId[1]].isOffScreen()){
            this.fishes[this.fishesId[1]].initialize();
            this.fishesId[1] = Math.round(Math.random()*1000) % 6;
            while(this.fishesId[0] === this.fishesId[1])
                this.fishesId[1] = Math.round(Math.random()*1000) % 6;            
        }
        //
        for (let i = 0; i < this.trashes.length; i++)
            this.trashes[i].update();
        //
        for (let i=0; i< 4; i++)
            this.dolphins[i].update();
        //
        for (let i = 0; i<4; i++)
            if(this.dolphins[i].state === DOLPHIN_FLOW)
                for (let j = 0; j < 10; j++)
                    if (!this.trashes[j].isCaptured 
                        && this.dolphins[i].mouthBB.isIntersect(this.trashes[j].bBox)){
                         this.trashes[j].isCaptured = true;
                         this.dolphins[i].trashId = j;
                         this.dolphins[i].setState(DOLPHIN_EAT);
                         this.onDolphinDie(this);
                         this.onTrashLoad(this);
                         //break;
                     }
        //
        for (let i = 0; i < 4; i++)
            if (this.dolphins[i].state === DOLPHIN_FLOW &&
                this.dolphins[i].bodyBB.isIntersect(Hug.bBox)) {
                this.dolphins[i].setState(DOLPHIN_DIE);
                this.onDolphinDie(this);
            }
        Ship.update();
    },
    //
    draw : function(){
        //
        //this.context.clearRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
        this.seaBk.draw(this.context);
        this.waves.drawBk(this.context);
        //
        for (let i =0; i< 2; i++)
            this.fishes[this.fishesId[i]].draw(this.context);
        //
        this.corrals.draw(this.context);
        //
        for (let i = 0; i < this.trashes.length; i++)
            this.trashes[i].draw(this.context);
        //   
        for (let i=0; i< 4; i++)
            this.dolphins[i].draw(this.context);
        //
        Ship.draw(this.context);
        //
        for (let i=0; i<20; i++)
            this.bubbles[i].draw(this.context);
        //       
        this.seaTop.draw(this.context);
        this.waves.drawTop(this.context);
        //
        switch(this.state){
            case GAME_PLAY :
                if(this.platePause.PosY > -420)
                    this.platePause.draw(this.context);
                break;
            case GAME_PAUSE :
                this.platePause.draw(this.context);
                break;
            case GAME_WON :
                this.plateWon.draw(this.context);
                break;
            case GAME_LOST :
                this.plateLost.draw(this.context);
                break;    
        }       
    },
    liveDolphins : function(){
        // проверка, есть ли хоть один плавающий дельфин
        for (let i =0; i< this.dolphins.length; i++)
            if(this.dolphins[i].state === DOLPHIN_FLOW)
                return true;
        return false;
    },       
    dieDolphins : function(){
        // проверка, все ли дельфины утонули
        for (let i =0; i< this.dolphins.length; i++)
            if(this.dolphins[i].state !== DOLPHIN_OUT)
                return false;
        return true;
    },   
    thereIsTrash : function(){
        for (let i = 0; i < this.trashes.length; i++)
            if(!this.trashes[i].isCaptured)
                return true;
        return false;
    },        
    isLost : function(){
        if(this.dieDolphins()){
            this.state = GAME_LOST;
            return true;            
        }
        else
            return false;
    },
    isWon : function(){
        if(this.liveDolphins() && !this.thereIsTrash()){
            this.state = GAME_WON;
            return true;
        }
        else
            return false;        
    },        
    //
    mainLoop : function(){
                
        this.update();
        this.draw();        
    },
    onKeyDown : function(event){
        switch(event.keyCode){
            case 37 : // left
                Screw.state = SCREW_LEFTWARD;
                break;
            case 39 : // right
                Screw.state = SCREW_RIGHTWARD;
                break;           
            case 38 : 
                Hug.moving = HUG_MOVE_UP;
                break;
            case 40 :
                Hug.moving = HUG_MOVE_DOWN;
                break;
            case 32 : 
                Hug.Catch();
                break;
            case 80 :
                if(Game.state === GAME_PLAY)
                    Game.state = GAME_PAUSE;
                else
                    if(Game.state === GAME_PAUSE)
                        Game.state = GAME_PLAY;
                break;                 
        }
    },
    //
    onKeyUp : function(event){
        switch(event.keyCode){
            case 37 :
                if(Screw.state === SCREW_LEFTWARD)
                    Screw.state = SCREW_STOPPED;
                break;
            case 39 :
                if(Screw.state === SCREW_RIGHTWARD)
                    Screw.state = SCREW_STOPPED;
                break;
            case 38 :
                if(Hug.moving === HUG_MOVE_UP)
                    Hug.moving = HUG_NO_MOVE;
                break;
            case 40 :
                if(Hug.moving === HUG_MOVE_DOWN)
                    Hug.moving = HUG_NO_MOVE;
                break;             
        }
    },        
    getLiveDolphinsCount: function () {
        let dolphinsCount = 0;
        for (let i = 0; i < this.dolphins.length; i++) {
            if (this.dolphins[i].state === DOLPHIN_FLOW)
                dolphinsCount++;
        }
        return dolphinsCount;
    },
    getTrashesInWater: function () {
        let trashesCount = 0;
        for (let i = 0; i < this.trashes.length; i++) {
            if (!this.trashes[i].isCaptured)
                trashesCount++;
        }
        return trashesCount;
    },
    startTime: null,
    isGameFinished: false,
    onGameFinish: function () {
        if (this.isGameFinished) return;

        let dolphinsCount = 0;
        for (let i = 0; i < this.dolphins.length; i++) {
            if (this.dolphins[i].state === DOLPHIN_FLOW)
                dolphinsCount++;
        }

        let totalSeconds = Math.floor((Date.now() - this.startTime) / 1000);


        this.onGameEndCallback({
            victory: dolphinsCount > 0,
            dolphins: dolphinsCount,
            seconds: totalSeconds
        });
        this.isGameFinished = true;
    },
    onGameEndCallback: function () { },
    start: function (_onGameEnd) {
        this.startTime = Date.now();
        this.timerId = setInterval('Game.mainLoop()', UPDATE_INTERVAL);
        window.addEventListener('keydown', Game.onKeyDown, true);
        window.addEventListener('keyup', Game.onKeyUp, true);
        this.onGameEndCallback = _onGameEnd;
    },
    onDolphinDie: function () { },
    onTrashLoad: function () { }
};

